flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java

```diff
@@ -82,18 +83,52 @@ synchronized void retainTaskManagers(List<String> activeTaskManagers) {
      */
     synchronized void retainJobs(List<String> activeJobs) {
         jobs.keySet().retainAll(activeJobs);
-        currentExecutionAttempts.keySet().retainAll(activeJobs);
+        representativeAttempts.keySet().retainAll(activeJobs);
     }
 
     public synchronized void updateCurrentExecutionAttempts(Collection<JobDetails> jobs) {
-        jobs.forEach(
-                job ->
-                        currentExecutionAttempts.put(
-                                job.getJobId().toString(), job.getCurrentExecutionAttempts()));
+        for (JobDetails job : jobs) {
+            String jobId = job.getJobId().toString();
+            Map<String, Map<Integer, List<Integer>>> currentExecutionAttempts =
+                    job.getCurrentExecutionAttempts();
+            Map<String, Map<Integer, Integer>> jobRepresentativeAttempts =
+                    new HashMap<>(currentExecutionAttempts.size());
+            for (Map.Entry<String, Map<Integer, List<Integer>>> vertexEntry :
+                    currentExecutionAttempts.entrySet()) {
+                String vertexId = vertexEntry.getKey();
+                Map<Integer, List<Integer>> subtaskAttempts = vertexEntry.getValue();
+                Map<Integer, Integer> vertexRepresentativeAttempts =
+                        new HashMap<>(subtaskAttempts.size());
+                for (Map.Entry<Integer, List<Integer>> e : subtaskAttempts.entrySet()) {
+                    int subtaskIndex = e.getKey();
+                    List<Integer> currentAttemptNumbers = e.getValue();
+                    retainSubtaskAttemptMetrics(
+                            jobId, vertexId, subtaskIndex, currentAttemptNumbers);
+                    if (!currentAttemptNumbers.isEmpty()) {
+                        vertexRepresentativeAttempts.put(
+                                subtaskIndex, currentAttemptNumbers.get(0));
+                    }
+                }
+                if (!vertexRepresentativeAttempts.isEmpty()) {
+                    jobRepresentativeAttempts.put(vertexId, vertexRepresentativeAttempts);
+                }
+            }
+            representativeAttempts.put(jobId, jobRepresentativeAttempts);
+        }
+    }
+
+    private void retainSubtaskAttemptMetrics(
+            String jobId, String vertexId, int subtaskIndex, List<Integer> currentAttempts) {
+        Map<Integer, ComponentMetricStore> attemptMetrics =
```

Comment:
We might move this method into SubtaskMetricStore so that we do not expose attemptMetrics  and modify it outside of its containing class.